### ISSUE TITLE
Can't get Ultisnips to work

### ISSUE THREAD

**Issue by xayden** [neovim/neovim#8085]
I ran these commands
```
$ sudo pip2 install --upgrade neovim
$ sudo pip2 install --user --upgrade neovim
$ sudo pip3.6 install --upgrade neovim
$ sudo pip3.6 install --user --upgrade neovim
```
and added the following to `init.vim`
```
let g:python_host_prog = 'user/bin/python'
let g:python3_host_prog = '/usr/local/bin/python3.6'

let g:loaded_python_provider = 0
let g:loaded_python3_provider = 0 
```
yet nothing seems to be working, especially Ultisnips, also this is the result of running 
`echo g:loaded_python_provider`, `echo g:loaded_python3_provider`, `echo has('python')` and 
`echo has('python3.6')`
```
0
0
0
0
```
When running `nvim` in terminal it prints out the following 
`UltiSnips requires py >= 2.7 or py3`
and running `:py import sys; print(sys.version_info)` prints out 
```
E117: Unknown function: provider#python#Call 
```

**Comment by mhinz** [comment]
Run `:checkhealth`.

I guess `/usr/bin/python -c 'import neovim'` reports `ImportError: No module named neovim`, right?

**Comment by justinmk** [comment]
> E117: Unknown function: provider#python#Call 

Looks like $VIMRUNTIME is invalid. As mhinz mentioned, run `:checkhealth` . 

And please don't ignore the issue template. You did not specify which version.

**Comment by xayden** [comment]
version
```
NVIM v0.2.2
Build type: RelWithDebInfo
LuaJIT 2.0.4
Compilation: /usr/bin/x86_64-linux-gnu-gcc -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -DDISABLE_LOG -Wdate-time -D_FORTIFY_SOURCE=2 -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_UNIBI_HAS_VAR_FROM -O2 -g -DMIN_LOG_LEVEL=3 -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/build/neovim-HJnLxJ/neovim-0.2.2/build/config -I/build/neovim-HJnLxJ/neovim-0.2.2/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim-HJnLxJ/neovim-0.2.2/build/src/nvim/auto -I/build/neovim-HJnLxJ/neovim-0.2.2/build/include
Compiled by pkg-vim-maintainers@lists.alioth.debian.org

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
```
running `:checkhealth` after removing `let g:loaded_python_provider=0` and `let g:loaded_python3_provider=0`
```
health#deoplete#check ========================================================================
## deoplete.nvim
  - OK: has("nvim") was successful
  - OK: exists("v:t_list") was successful
  - OK: has("python3") was successful
  - INFO: If you're still having problems, try the following commands:
    $ export NVIM_PYTHON_LOG_FILE=/tmp/log
    $ export NVIM_PYTHON_LOG_LEVEL=DEBUG
    $ nvim
    $ cat /tmp/log_{PID}
    and then create an issue on github

health#nvim#check
========================================================================
## Configuration
  - OK: no issues found

## Performance
  - OK: Build type: RelWithDebInfo

## Remote Plugins
  - WARNING: "deoplete.nvim" is not registered.
  - WARNING: Out of date
    - ADVICE:
      - Run `:UpdateRemotePlugins`

## terminal
  - INFO: key_backspace (kbs) terminfo entry: key_backspace=\177
  - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~
  - INFO: $XTERM_VERSION=''
  - INFO: $TERM_PROGRAM=''
  - INFO: $COLORTERM=''
  - INFO: $SSH_TTY=''

health#provider#check
========================================================================
## Clipboard (optional)
  - OK: Clipboard tool found: xsel

## Python 2 provider (optional)
  - INFO: Using: g:python_host_prog = "user/bin/python"
  - INFO: Executable: /usr/bin/python
  - ERROR: Command error (job=5): '/usr/bin/python' '-c' 'import neovim; print(neovim.__file__)'
    Output: Traceback (most recent call last):  File "<string>", line 1, in <module>ImportError: No module named neovim
  - INFO: Python2 version: 2.7.12
  - INFO: python-neovim version: unable to load neovim Python module
  - ERROR: Neovim Python client is not installed.
    Error: unable to load neovim Python module
    - ADVICE:
      - Run in shell: pip2 install neovim

## Python 3 provider (optional)
  - INFO: Using: g:python3_host_prog = "/usr/local/bin/python3.6"
  - INFO: Executable: /usr/local/bin/python3.6
  - ERROR: Command error (job=8): '/usr/local/bin/python3.6' '-c' 'import neovim; print(neovim.__file__)'
    Output: Traceback (most recent call last):  File "<string>", line 1, in <module>ModuleNotFoundError: No module named 'neovim'
  - INFO: Python3 version: 3.6.3
  - INFO: python3.6-neovim version: unable to load neovim Python module
  - ERROR: Neovim Python client is not installed.
    Error: unable to load neovim Python module
    - ADVICE:
      - Run in shell: pip3 install neovim

## Ruby provider (optional)
  - INFO: Ruby: ruby 2.3.1p112 (2016-04-26) [x86_64-linux-gnu]
  - WARNING: Missing "neovim" gem.
    - ADVICE:
      - Run in shell: gem install neovim
      - Is the gem bin directory in $PATH? Check `gem environment`.
      - If you are using rvm/rbenv/chruby, try "rehashing".

## Node provider (optional)
  - INFO: Node: v8.9.4
  - INFO: Host: /usr/bin/neovim-node-host
  - OK: Latest "neovim" npm is installed: 3.5.2
```

**Comment by justinmk** [comment]
>   - ERROR: Command error (job=8): '/usr/local/bin/python3.6' '-c' 'import neovim; print(neovim.__file__)'

`:checkhealth` is saying very plainly what it tried to do and what didn't work. Does this work in your shell?

     /usr/local/bin/python3.6 -c 'import neovim'

If it doesn't work, then whatever pip command you used to install, was not picked up by `python3.6`. Why did you set `g:python3_host_prog=/usr/local/bin/python3.6` ?

These are python questions, not something we can handle on the issue tracker.

**Comment by mhinz** [comment]
Did you follow my steps? Also..

```
INFO: Using: g:python_host_prog = "user/bin/python"
```

I guess that's not a valid path.

**Comment by xayden** [comment]
Python 2.7 support doesn't work at all, no matter how many times I uninstall and reinstall running 
`python -c 'import neovim'` always prints `ImportError: No module named neovim`
but python3 works fine.

**Comment by mhinz** [comment]
@xayden 

If you can't figure it out, visit our [Gitter channel](https://gitter.im/neovim/neovim) and we'll find a way. But this tracker is meant for development.

**Comment by jdhao** [comment]
I think I have found the cause of this problem. I will elaborate so others with the same issue can solve similar issues. 

# My environent:

+ Platform: CentOS  7 and Windows 10
+ Neovim: 0.3.2
+ Python: 3.6.5 (provided by Anaconda)

# How did the issue arise and how I solved it
Initially, in order to make auto-completion work for Python. I installed deoplete, deoplete-jedi (which requires installation of neovim python client and jedi). The auto-completion works fine. The python neovim package version is 0.2.6, IIRC. 

After some time, I run `:checkhealth` inside neovim. It told me that the neovim python client is outdated and there is a new version 0.3.1 out there. 

I upgrade neovim python client with `pip install -U neovim` and `pip list|grep neovim` shows:

> neovim                             0.3.1

This is the start of endless struggle...

After that, when I start nvim. It complains:

> deoplete failed to load. try the updateremoteplugins command and restart neovim

I used the comamnd `:UpdateRemotePlugins`, and see the following errors:

> UltiSnips requires py >= 2.7 or py3
> 

I add the following setting in my `init.vim`:

```
let g:python3_host_prog='/home/haojiedong/tools/anaconda3/bin/python'
```

The above warning message disappears. Each time I open nvim, I will see the error message:

> [deoplete] deoplete failed to load. Try the :UpdateRemotePlugins command and restart Neovim. See also :checkhealth.


I run the `:UpdateRemotePlugins` command in nvim and get the following error message:

```
function remote#host#UpdateRemotePlugins[6]..<SNR>36_RegistrationCommands[15]..remote#host#Require[10]..provider#pythonx#Require[13]..provider#Poll, line 4
Vim(if):ch 3 was closed by the client
function remote#host#UpdateRemotePlugins[6]..<SNR>36_RegistrationCommands[15]..remote#host#Require[10]..provider#pythonx#Require[13]..provider#Poll, line 14
Failed to load python3 host. You can try to see what happened by starting nvim with $NVIM_PYTHON_LOG_FILE set and opening the generated log file. Also, the host stderr is available in messages.
remote/host: generated rplugin manifest: /home/haojiedong/.local/share/nvim/rplugin.vim
```

I may also get the error message:

```
function remote#host#UpdateRemotePlugins[6]..<SNR>36_RegistrationCommands[15]..remote#host#Require[10]..provider#pythonx#Require[13]..provider#Poll, line 4
Vim(if):ch 3 was closed by the client
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'neovim'
function remote#host#UpdateRemotePlugins[6]..<SNR>36_RegistrationCommands[15]..remote#host#Require[10]..provider#pythonx#Require[13]..provider#Poll, line 14
Failed to load python3 host. You can try to see what happened by starting nvim with $NVIM_PYTHON_LOG_FILE set and opening the generated log file. Also, the host stderr is available in messages.
remote/host: generated rplugin manifest: /home/haojiedong/.local/share/nvim/rplugin.vim
```

I tried to uninstall neovim and reinstall:

```
pip uninstall neovim
pip install neovim
# or pip install --no-cache-dir neovim
```
The error persists.

Why is neovim package not found but pip reports that neovim is installed? I go to the Python package directory (`$ANACONDA_ROOT/lib/python3.6/site-packages`). `ls |grep neovim` shows:

>205:neovim-0.3.1.dist-info/

So the neovim package is actually not installed! There is only wheel info. This is strange and I do not know why.

I tried to uninstall neovim and all its dependencies and do a complete re-install:

```
pip uninstall -y neovim pynvim msgpack greenlet
pip install neovim
```

Now, `ls|grep neovim` command in Anaconda Python site package directory shows:

> 205:neovim/
> 206:neovim-0.3.1.dist-info/
> 

So the neovim Python client is successfully installed. Now open neovim and use issue `UpdateRemotePlugins` command, you will see:

```
remote/host: python3 host registered plugins ['deoplete']
remote/host: generated rplugin manifest: /home/haojiedong/.local/share/nvim/rplugin.vim
```

Restart Neovim and the error message will disappear. You will be able to use autocompletion again.

The only question is: why didn't `pip install -U neovim` install update neovim package correctly?

**Comment by mazzma12** [comment]
```
pip uninstall -y neovim pynvim msgpack greenlet
pip install neovim
```
This fixed it for me on Ubuntu 16.04. The problem came from neovim dependencies as I reinstalled neovim already without success.

**Comment by justinmk** [comment]
@mazzma12 That's not correct. The package was renamed to `pynvim` as mentioned in the FAQ and Following-HEAD wiki pages.

- https://github.com/neovim/neovim/wiki/Following-HEAD#20181118
- https://github.com/neovim/neovim/wiki/FAQ#python-support-isnt-working

See either page for correct instructions.
